pkg/deploy/predeploy.go

```diff
@@ -439,25 +463,91 @@ func (d *deployer) createSecret(ctx context.Context, kv keyvault.Manager, secret
 	})
 }
 
-func (d *deployer) ensureSecretKey(ctx context.Context, kv keyvault.Manager, secretName string) error {
+func (d *deployer) ensureSecretKey(ctx context.Context, kv keyvault.Manager, secretName string) (isNew bool, err error) {
 	existingSecrets, err := kv.GetSecrets(ctx)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	for _, secret := range existingSecrets {
 		if filepath.Base(*secret.ID) == secretName {
-			return nil
+			return false, nil
 		}
 	}
 
 	key, err := rsa.GenerateKey(rand.Reader, 2048)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	d.log.Infof("setting %s", secretName)
-	return kv.SetSecret(ctx, secretName, azkeyvault.SecretSetParameters{
+	return true, kv.SetSecret(ctx, secretName, azkeyvault.SecretSetParameters{
 		Value: to.StringPtr(base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PrivateKey(key))),
 	})
 }
+
+func (d *deployer) restartOldScalesets(ctx context.Context, resourceGroupName string) error {
+	d.log.Print("restarting old scalesets")
+	scalesets, err := d.vmss.List(ctx, resourceGroupName)
+	if err != nil {
+		return err
+	}
+
+	for _, vmss := range scalesets {
+		err = d.restartOldScaleset(ctx, *vmss.Name, resourceGroupName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *deployer) restartOldScaleset(ctx context.Context, vmssName string, resourceGroupName string) error {
+	var restartScript string
+	var waitForReadiness func(ctx context.Context, vmssName string) error
+	switch {
+	case strings.HasPrefix(vmssName, gatewayVMSSPrefix):
+		restartScript = gatewayRestartScript
+		waitForReadiness = d.gatewayWaitForReadiness
+	case strings.HasPrefix(vmssName, rpVMSSPrefix):
+		restartScript = rpRestartScript
+		waitForReadiness = d.rpWaitForReadiness
+	default:
+		return nil
```

Comment:
I am not sure if we want to stop the RP rollout for a VMSS that isn't expected to be present in the resource group by a check here, this function is supposed to deal with restarting the gateway and rp VMs or skip otherwise.

Reply:
We may at least want to return an error like `provided vmss does not match RP or gateway prefix` here

Reply:
Added an error return as suggested.

Reply:
Also added test cases for entire predeploy functionality.